### ISSUE TITLE
al_transform_coordinates_3d_projection usage confusion

### ISSUE THREAD

**Issue by CapSel** [liballeg/allegro5#905]
https://gist.github.com/CapSel/9388b63396ddd14a2f44d20bdecbfc6b

In the gist above I tried to demonstrate my problem.

I wanted to calculate screen (2d) coordinates for center of the grey square so I could detect when mouse is pointed at it. Unfortunately position of green and red squares does not correspond to center of grey square - they are really far from it.

What have I done wrong? Could it be Allegro problem?

**Comment by allefant** [comment]
Thanks, the issue is with Allegro I would say.

al_transform_coordinates_3d_projective was added to fix this issue: https://github.com/liballeg/allegro5/issues/863

It calls al_transform_coordinates_4d and then divides the resulting coordinates by w. However that will return normalized coordinates. To get the actual pixel positions we need to scale and translate them. The question is now if the function has use as-is (but we need to update the documentation), or if we should just fix it to be useful (and return pixel coordinates).

**Comment by allefant** [comment]
Basically al_orthographic_transform and al_perspective_transform are not very well documented currently. The former just says "Combines the given transformation with an orthographic transformation which maps the screen rectangle to the given left/top and right/bottom coordinates.".

What that "maps" really means is, it creates a transformation which when applied to an x/y/z coordinate, will result in a coordinate that is inside a cube from -1/-1/-1 to 1/1/1 if x/y/z was inside the cube from left/top/near to right/bottom/far and else outside. That in itself is rather useless. However if we pass that transformation to al_use_projection_transform we pass it on to OpenGL/DirectX and it just so happens that this is the internal coordinate system expected by those. There probably is zero use cases for the result of al_orthographic/perspective_transform other than passing it to al_use_projection_transform.

al_transform_coordinates_3d_projective now will return those same coordinates inside of -1/-1/-1 to 1/1/1 for any x/y/z passed to it. However this time there is nothing else to pass it to. At the same time we simply do not have the information how to interpret the coordinates either because we do not know in which display al_use_projection_transform was called with that same transform. (We could use the currently active display though.)

So some suggested options I would have:

- A. Modify al_transform_coordinates_3d_projective to use the current display in addition to the passed transformation to return actual pixel positions.
- B. Add a new function which works like above, maybe al_transform_coordinates_to_current_display
- C. Like above but do not pass the transformation, instead just use whatever was passed to al_use_projection_transform.
- D. Like above but explicitly pass both an ALLEGRO_DISPLAY and an ALLEGRO_TRANSFORM.
- E. Like B./C./D. but pass the result of the current al_transform_coordinates_3d_projective to the new function, so you would always have two function calls (just like you always call al_perspective_transform followed by al_use_projective_transform)

**Comment by SiegeLord** [comment]
It's not particularly hard to to do the scaling from the unit cube to the target bitmap though. I think an example will suffice.

```c
ALLEGRO_TRANSFORM t2;
al_copy_transform(&t2, al_get_current_transform());
al_compose_transform(&t2, al_get_current_projection_transform());

ALLEGRO_TRANSFORM t3;
al_identity_transform(&t3);
al_scale_transform(&t3, 0.5, -0.5);
al_translate_transform(&t3, 0.5, 0.5);
al_scale_transform(&t3, al_get_bitmap_width(al_get_target_bitmap()),
                   al_get_bitmap_height(al_get_target_bitmap()));

al_transform_coordinates_3d_projective(&t2, &x, &y, &z);
al_transform_coordinates(&t3, &x, &y);
// x, y now contain pixel coordinates
```

**Comment by allefant** [comment]
Yes, or if al_transform_coordinates_3d_projective returns x/y, just do:

x = w / 2 + x * w / 2
y = h / 2 - y * h / 2

I'd be fine with adding that to the documentation and not making code changes!

**Comment by CapSel** [comment]
I used @allefant 's solution and it works fine.

What do you think about adding another parameter to al_transform_coordinates_3d_projective - screen size (or two ints width, height)?
